### PR TITLE
Use function pointers in crossgen2

### PR DESCRIPTION
**Pull request by jkotas**

### CODE REVIEW

**Comment by jkotas** [comment]
cc @dotnet/crossgen-contrib

**Review by MichalStrehovsky — APPROVED** [review]
Neat! I've been thinking about doing this ever since we picked up a Roslyn that has function pointer support in all places that compile these files! This, and the `AddrOf` intrinsics.

Thank you!

Out of curiosity - did you see a throughput improvement? Just the simplification is totally worth it, but TP improvement would be a nice bonus.

**Comment by mangod9** [comment]
yeah this is great, hoping there is some perf gain too.

File: src/coreclr/src/tools/Common/JitInterface/ThunkGenerator/Program.cs [hunk 1]
```diff
@@ -249,80 +264,23 @@ unsafe partial class CorInfoImpl
...
                 tr.Write(@")
         {
             var _this = GetThis(thisHandle);
             try
             {
 ");
-                bool isVoid = decl.ReturnAsParm || decl.ReturnType.ManagedTypeName == "void";
-                tr.Write("                " + (isVoid ? "" : "return ") + "_this." + decl.FunctionName + "(");
+                bool isVoid = decl.ReturnType.ManagedTypeName == "void";
+                tr.Write($"                " + (isVoid ? "" : "return ") + $"_this.{decl.FunctionName}(");
```
Comment (am11) [inline src/coreclr/src/tools/Common/JitInterface/ThunkGenerator/Program.cs:282]: nit: fold in single interpolation
```suggestion
                tr.Write($"                {(isVoid ? "" : "return ")}_this.{decl.FunctionName}(");
```

**Comment by jkotas** [comment]
> the AddrOf intrinsics.

I am waiting with cleaning up the AddrOf intrinsics and converting delegates under src/libraries for the final version of Roslyn. The version of Roslyn that we got does not have the proper type inference for `UnmanagedCallersOnly`. Once we get that version, we will need to add `unmanaged` to all places that take address of `UnmanagedCallersOnly` to fix compile errors and pass it around. It would be annoying to do that in hundred different files.

> did you see a throughput improvement? 

I did not bother with measuring it since it will be in the noise range. If I remember correctly from looking at the profiles, the  marshalled delegates did not show as significant overhead.

**Pull request merged**